neo/Consensus/ConsensusService.cs

```diff
@@ -220,17 +220,33 @@ private void OnChangeViewReceived(ConsensusPayload payload, ChangeView message)
 
         private void OnCommitReceived(ConsensusPayload payload, Commit commit)
         {
-            if (context.CommitPayloads[payload.ValidatorIndex] != null) return;
-            Log($"{nameof(OnCommitReceived)}: height={payload.BlockIndex} view={commit.ViewNumber} index={payload.ValidatorIndex}");
-            byte[] hashData = context.MakeHeader()?.GetHashData();
-            if (hashData == null)
+            ref ConsensusPayload existingCommitPayload = ref context.CommitPayloads[payload.ValidatorIndex];
+            if (commit.ViewNumber == context.ViewNumber)
             {
-                context.CommitPayloads[payload.ValidatorIndex] = payload;
+                Log($"{nameof(OnCommitReceived)}: height={payload.BlockIndex} view={commit.ViewNumber} index={payload.ValidatorIndex}");
+
+                byte[] hashData = context.MakeHeader()?.GetHashData();
+                if (hashData == null)
+                {
+                    existingCommitPayload = payload;
+                }
+                else if (Crypto.Default.VerifySignature(hashData, commit.Signature,
+                    context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+                {
+                    existingCommitPayload = payload;
+                    CheckCommits();
+                }
+                return;
             }
-            else if (Crypto.Default.VerifySignature(hashData, commit.Signature, context.Validators[payload.ValidatorIndex].EncodePoint(false)))
+            // Receiving commit from another view
+            Log($"{nameof(OnCommitReceived)}: record commit for different view={commit.ViewNumber} index={payload.ValidatorIndex} height={payload.BlockIndex}");
+            if (existingCommitPayload == null)
+                existingCommitPayload = payload;
+            else
```

Comment:
I do not agree in modifying the commit payload that you already receive from a payload. I prefer to remove this else. 
Following our logic, there is no change to receive a different commit for that payload, it should not happen.
A warning could be logged in such case.

Reply:
Let's talk offline.

Reply:
Ok, I agree. Let's remove the else.